### PR TITLE
pump,pkg/util: Refactor and add more unit tests

### PR DESCRIPTION
**Pull request by suzaku**
### What problem does this PR solve? 

[Jira Issue](https://internal.pingcap.net/jira/browse/TOOL-1056) 

### What is changed and how it works?

Improve tests coverage: 44.2% -> 54.6%

### Check List 

Tests 

 - Unit test

Code changes

Side effects

Related changes

### CODE REVIEW

File: pump/server.go [hunk 1]
```diff
@@ -828,3 +782,33 @@ func (s *Server) Close() {
...
+			}
+		}
+		return nil
+	}
+}
+
+func listen(network, addr string) (net.Listener, error) {
+	URL, err := url.Parse(addr)
+	if err != nil {
+		return nil, errors.Annotatef(err, "invalid listening socket addr (%s)", URL)
```
Comment (july2993) [inline pump/server.go]: s/URL/add ?

File: pump/server.go [hunk 2]
```diff
@@ -692,77 +685,38 @@ func (s *Server) waitSafeToOffline(ctx context.Context) error {
 	var fakeBinlog *pb.Binlog
 	var err error
 
-	// write a fake binlog
-	for {
+	err = util.TryUntilSuccess(ctx, time.Second, "Failed to write fake binlog", func() error {
 		fakeBinlog, err = s.writeFakeBinlog()
```
Comment (july2993) [inline pump/server.go]: i think this `err` will capture the `err` at L686?
will it clear to declare a new `err` inside the annoy func?

File: pump/server.go [hunk 3]
```diff
@@ -692,77 +685,38 @@ func (s *Server) waitSafeToOffline(ctx context.Context) error {
...
 
-	// check drainer has consume fake binlog we just write
 	for {
 		select {
 		case <-time.After(time.Second):
-			pumpNode := s.node.(*pumpNode)
-			drainers, err := pumpNode.Nodes(ctx, "drainers")
+			minSavedTSO, err := s.getSaveGCTSOForDrainers()
 			if err != nil {
 				log.Error("get nodes failed", zap.Error(err))
```
Comment (july2993) [inline pump/server.go]: update  log too

File: pump/server.go [hunk 4]
```diff
@@ -692,77 +685,38 @@ func (s *Server) waitSafeToOffline(ctx context.Context) error {
...
-	log.Debug("start to check offline safe for drainers")
+	log.Debug("Start waiting until all drainers have consumed the last fake binlog")
 
-	// check drainer has consume fake binlog we just write
 	for {
 		select {
 		case <-time.After(time.Second):
-			pumpNode := s.node.(*pumpNode)
-			drainers, err := pumpNode.Nodes(ctx, "drainers")
+			minSavedTSO, err := s.getSaveGCTSOForDrainers()
```
Comment (july2993) [inline pump/server.go]: s/Save/Safe , i think it's a typo

**Comment by july2993** [comment]
@amyangfei PTAL

File: pkg/util/util.go [hunk 5]
```diff
@@ -151,6 +152,27 @@ func RetryOnError(retryCount int, sleepTime time.Duration, errStr string, fn fun
 	return errors.Trace(err)
 }
 
+// TryUntilSuccess retries the given function until error is nil or the context is done,
+// waiting for `waitInterval` time between retries.
+func TryUntilSuccess(ctx context.Context, waitInterval time.Duration, errMsg string, fn func() error) error {
+	for {
+		if err := fn(); err != nil {
+			if errMsg != "" {
```
[the remark this answers is not in the document]
Comment (suzaku) [reply pkg/util/util.go]: `errMsg` is only empty when the user explicitly passes `""`, it is a way for the user to turn off logging.

File: pump/server.go [hunk 6]
```diff
@@ -502,19 +495,19 @@ func (s *Server) gcBinlogFile() {
...
-			safeTSO, err := s.getSaveGCTSOForDrainers()
+			safeTSO, err := s.getSafeGCTSOForDrainers()
 			if err != nil {
 				log.Warn("get save gc tso for drainers failed", zap.Error(err))
 				continue
 			}
 			log.Info("get safe ts for drainers success", zap.Int64("ts", safeTSO))
 
-			millisecond := time.Now().Add(-s.gcDuration).UnixNano() / 1000 / 1000
+			millisecond := time.Now().Add(-s.gcDuration).UnixNano() / 1e6
```
[the remark this answers is not in the document]
Comment (suzaku) [reply pump/server.go]: Just a more succinct way to write `/ 1000 / 1000`.

**Comment by WangXiangUSTC** [comment]
LGTM

File: pump/server.go (hunk 6 above)
[the remark this answers is not in the document]
Comment (WangXiangUSTC) [reply pump/server.go]: but I think / 1000 / 1000 is more clear

[the remark this answers is not in the document]
Comment (suzaku) [reply pump/server.go]: OK, I can change it back.

Comment (suzaku) [inline pump/server.go]:
```suggestion
			millisecond := time.Now().Add(-s.gcDuration).UnixNano() / 1000 / 1000
```

**Comment by suzaku** [comment]
@WangXiangUSTC  Fixed.

**Pull request merged**